File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Set<Character> s1=new HashSet<>();
        int l=0,r=0,m=0;
        while(r<s.length()){
            while(s1.contains(s.charAt(r))){
                s1.remove(s.charAt(l));
                l++;
            }
            s1.add(s.charAt(r));
            m=Math.max(m,r-l+1);
            r++;
        }
return m;
    }
}

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        l,r,maxLen=0,0,0
        lastSeen={}
        while r<len(s):
            if s[r] in lastSeen and lastSeen[s[r]]>=l:
                l=lastSeen[s[r]]+1
            lastSeen[s[r]]=r
            maxLen=max(maxLen,r-l+1)
            r+=1
        return maxLen